boot/freeldr/bootsect/ntfs.S

```diff
@@ -0,0 +1,804 @@
+
+ReadSectorsCHSLoop:
+    pushad
+    xor  edx, edx
+    movzx ecx, word ptr [SectorsPerTrack]
+    div  ecx                                    // Divide logical by SectorsPerTrack
+    inc  dl                                     // Sectors numbering starts at 1 not 0
+    mov  cl, dl                                 // Sector in CL
+    mov  edx, eax
+    shr  edx, 16
+    div  word ptr [NumberOfHeads]               // Divide logical by number of heads
+    mov  dh, dl                                 // Head in DH
+    mov  dl, byte ptr [BootDrive]               // Drive number in DL
+    mov  ch, al                                 // Cylinder in CX
+    ror  ah, 1                                  // Low 8 bits of cylinder in CH, high 2 bits
+    ror  ah, 1                                  //  in CL shifted to bits 6 & 7
+    or   cl, ah                                 // Or with sector number
+    mov  ax, HEX(0201)
+    int  HEX(13)    // DISK - READ SECTORS INTO MEMORY
+                     // AL = number of sectors to read, CH = track, CL = sector
+                     // DH = head, DL = drive, ES:BX -> buffer to fill
+                     // Return: CF set on error, AH = status (see AH=01h), AL = number of sectors read
+
+    jc   PrintDiskError                         // If the read failed then abort
+
+    popad
+
+    inc  eax                                    // Increment Sector to Read
+
+    mov  dx, es
+    add  dx, 32                                 // Increment read buffer for next sector
+    mov  es, dx
+
+    loop ReadSectorsCHSLoop                     // Read next sector
+
+    pop es
+    ret
+
+// Displays a disk error message
+// And reboots
+PrintDiskError:
+    mov  si, offset msgDiskError        // Bad boot disk message
+    call PutChars                       // Display it
+
+    jmp  Reboot
+
+// Displays a file system error message
+// And reboots
+PrintFileSystemError:
+    mov  si, offset msgFileSystemError  // FreeLdr not found message
+    call PutChars                       // Display it
+
+Reboot:
+    mov  si, offset msgAnyKey           // Press any key message
+    call PutChars                       // Display it
+    xor  ax, ax
+    int  HEX(16)                        // Wait for a keypress
+    int  HEX(19)                        // Reboot
+
+PutChars:
+    lodsb
+    or   al, al
+    jz   short Done
+    mov  ah, HEX(0e)
+    mov  bx, 7
+    int  HEX(10)
+    jmp  short PutChars
+Done:
+    ret
+
+msgDiskError:
+    .ascii "Disk error", CR, LF, NUL
+msgFileSystemError:
+    .ascii "File system error", CR, LF, NUL
+msgAnyKey:
+    .ascii "Press any key to restart", CR, LF, NUL
+
+SectsPerMFT:
+    .word 0
+    
+MFTStartSector:
+    .long 0
+
+BiosCHSDriveSize:
+    .long 0
+
+.org 509 // Pad to 509 bytes
+BootPartition:
+    .byte 1
+
+BootFlagSignature:
+    .word HEX(0aa55)    // BootSector signature
+
```

Comment:
Please add an information block along the lines of what is done in the FAT32 bootsector:
https://git.reactos.org/?p=reactos.git;a=blob;f=boot/freeldr/bootsect/fat32.S;h=ff9c40af1211a7f7112b0a74d2bc7cdd876a8a0d;hb=HEAD#l317

Also, please check whether on Windows the NTFS boot sector is stored on disk contiguously or not. If not, please also add a comment similar to the one written in the FAT32 bootsector, concerning the second part that is (in FAT32 case) at sector 14.

Reply:
Boot sector is stored on disk contiguously on my Windows partitions: it's represented by the $Boot file (inode number 7), and uses the two first clusters in the first partition as marked in the run lists. Also, linux-ntfs documentation states this file cannot be relocated: https://flatcap.org/linux-ntfs/ntfs/files/boot.html.

I'll add what I wrote here as comment for the boot sector